src/rmkit/ui/events.cpy

```diff
@@ -0,0 +1,185 @@
+#include <cmath>
+
+#include "../input/events.h"
+#include "../util/signals.h"
+#include "timer.h"
+
+namespace ui:
+  PLS_DEFINE_SIGNAL(MOUSE_EVENT, input::SynMotionEvent)
+  class MOUSE_EVENTS:
+    public:
+    MOUSE_EVENT enter
+    MOUSE_EVENT leave
+    MOUSE_EVENT down
+    MOUSE_EVENT up
+    MOUSE_EVENT click
+    MOUSE_EVENT hover
+    MOUSE_EVENT move
+  ;
+
+  PLS_DEFINE_SIGNAL(KEY_EVENT, input::SynKeyEvent)
+  class KEY_EVENTS:
+    public:
+    KEY_EVENT pressed
+  ;
+
+  class GESTURE_EVENTS:
+    public:
+    MOUSE_EVENT long_press
+    MOUSE_EVENT single_click
+    MOUSE_EVENT double_click
+    MOUSE_EVENT drag_start
+    MOUSE_EVENT dragging
+    MOUSE_EVENT drag_end
+
+    // The threshold area that determines if this is a touch-based gesture
+    // (e.g. long_press, double_click) or a motion-based gesture (e.g. drag).
+    int touch_slop_size = 50
+
+    // Once dragging, only submit events when the x or y delta > this step.
+    // This limits the number of dragging events.
+    int dragging_step_size = 25
+
+    // Length of time (in ms) between a down event and a long_press
+    long long_press_timeout = 500
+
+    // Maximum gap (in ms) between clicks for a gesture to be considered a
+    // double_click. If there are any double_click handlers registered, this
+    // delay must pass without another down event before single_click will be
+    // triggered (otherwise single_click is triggered immediately on up).
+    long double_click_timeout = 250
+
+    // State machine
```

Comment:
Once I added double_click it felt like the logic was getting scattered around too much. I think it's easier to follow the logic when it's framed as a state machine.

Reply:
yes, definitely makes sense